app/services/onboarding_csv_parser.rb

```diff
@@ -0,0 +1,101 @@
+# frozen_string_literal: true
+
+require 'csv'
+
+# Error raised if an object in the CSV is not found in the db
+class ItemNotFoundError < StandardError; end
+
+#--------------------------
+#
+# @class OnboardingCsvParser
+#
+# @desc Responsibility: Parse the CSV IO given. Create
+#   objects as needed if they don't already exist.
+#
+# @author Ashley Engelund (ashley.engelund@gmail.com  weedySeaDragon @ github)
+# @date   11/01/20
+#
+#--------------------------
+#
+class OnboardingCsvParser
+  DATE_FORMAT = '%m-%d-%Y'
+  CONVERTERS = %i[numeric date].freeze
+
+  # ----------------------------------------------------------------------
+
+  def self.parse(data_string)
+    csv_rows = CSV.parse(data_string,
+                         headers: true,
+                         return_headers: false,
+                         unconverted_fields: %i[business_zip_code],
+                         converters: CONVERTERS)
+
+    csv_rows.each do |row|
+      business = get_business(row)
+      import_child_and_related(row, business)
+    end
+  end
+
+  # Get the business from the db. Raise error if the business is not found.
+  # @raise ItemNotFoundError
+  def self.get_business(row)
+    zipcode = Zipcode.find_by(code: row['business_zip_code'].to_s)
+    raise_not_found_error("Zipcode #{row['business_zip_code']}") unless zipcode
+
+    biz_name = row['business_name'].strip
+    business = Business.find_by(name: biz_name, zipcode: zipcode)
+    raise_not_found_error("Business '#{biz_name}'") unless business
+
+    business
+  end
+
+  # Create the child if it doesn't already exist and then create the ChildApproval for it
+  def self.import_child_and_related(row, business)
+    full_name = "#{row['first_name'].strip} #{row['last_name'].strip}"
+    dob = row['date_of_birth']
+    return if Child.find_by(full_name: full_name, date_of_birth: dob, business: business)
+
+    new_child = Child.create!(full_name: full_name,
+                              date_of_birth: dob,
+                              business: business,
+                              active: true,
+                              approvals: [get_approval(row)])
+
+    create_child_approval(new_child, new_child.approvals.first)
+  end
+
+  def self.get_approval(row)
+    approval_data = { case_number: row['case_number'].to_s.strip }
+    approval_data[:effective_on] = row['effective_on'] if row['effective_on'].present?
+    approval_data[:expires_on] = row['expires_on'] if row['expires_on'].present?
+    approval = Approval.find_by(approval_data)
+    raise_not_found_error("Approval case number: #{row['case_number']} effective on: #{row['effective_on']} expires on: #{row['expires_on']}") unless approval
+
+    approval
+  end
+
+  def self.create_child_approval(child, approval)
```

Comment:
Because this info will later come in via API, I'd rater this kind of logic, creating the approval (if it doesn't already exist), associating the child_approval with the subsidy_rule and rate_types, etc., should live in a service class that can be called either by the CSV parser or the child's create and update controller methods.  I'd love to see this parser just turn the csv into a JSON object just like the API would send, and then have the CSV Reader call the service on the result to associate it with these complicated associations.

Reply:
Got it. 
Makes sense.